util/dvsim/SgeLauncher.py

```diff
@@ -0,0 +1,174 @@
+            del exports['MAKEFLAGS']
+
+        self._dump_env_vars(exports)
+
+        try:
+            f = open(self.deploy.get_log_path(),
+                     "w",
+                     encoding="UTF-8",
+                     errors="surrogateescape")
+            f.write("[Executing]:\n{}\n\n".format(self.deploy.cmd))
+            f.flush()
+            # ---------- prepare SGE job struct -----
+            sgeJob = SGE.qsubOptions()
+            sgeJob.args.N = 'VCS_RUN_' + str(pid)  # Name of Grid Engine job
+            if "build.log" in self.deploy.get_log_path():
+                sgeJob.args.N = 'VCS_BUILD_' + str(pid)  # Name of Grid Engine job
+
+            job_name = sgeJob.args.N
+            sgeJob.args.t = '0'      # Define an array job with 20 subjobs
+            sgeJob.args.slot = '1'      # Define num of slot
+            sgeJob.args.sync = 'y'      # wait for the job to complete before exiting
+            sgeJob.args.q = 'vcs_q'   # Define the sge queue name
+            sgeJob.args.l = 'mf=20G'  # memory req,request the given resources
+            sgeJob.args.p = '0'      # Set priority to 0
+            # pecifies a range of priorities from -1023 to 1024.
+            # The higher the number, the higher the priority.
+            # The default priority for jobs is zero
+            sgeJob.args.command = '"' + self.deploy.cmd + '"'
+            sgeJob.args.b = 'y'      # This is a binary file
+            sgeJob.args.o = self.deploy.get_log_path() + '.sge'
+            cmd = str(sgeJob.execute(mode = 'echo'))
+            # ---------------
+            self.process = subprocess.Popen(shlex.split(cmd),
+                                            bufsize=4096,
+                                            universal_newlines=True,
+                                            stdout=f,
+                                            stderr=f,
+                                            env=exports)
+        except subprocess.SubprocessError as e:
+            raise LauncherError('IO Error: {}\nSee {}'.format(
+                e, self.deploy.get_log_path()))
+        finally:
+            self._close_process()
+
+        self._link_odir("D")
+        f.close()
+    def poll(self):
+        '''Check status of the running process
+
+        This returns 'D', 'P' or 'F'. If 'D', the job is still running. If 'P',
+        the job finished successfully. If 'F', the job finished with an error.
+
+        This function must only be called after running self.dispatch_cmd() and
+        must not be called again once it has returned 'P' or 'F'.
+        '''
+
+        assert self.process is not None
+        if self.process.poll() is None:
+            return 'D'
+        # -------------------------------------
+        # copy SGE jobb results to log file
+        if os.path.exists(self.deploy.get_log_path() + '.sge'):
+
+            file1 = open(self.deploy.get_log_path() + '.sge', 'r')
+            Lines = file1.readlines()
```

Comment:
Lines -> lines (naming guidelines). But it may be better to just do something like f.write(file1.read()), meaning don't bother to do per-line processing.

Reply:
fixed